functions/concepts/Retry/Function.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START functions_tips_retry]
+using CloudNative.CloudEvents;
+using Google.Cloud.Functions.Framework;
+using Google.Events.SystemTextJson.Cloud.PubSub.V1;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Retry
+{
+    public class Function : ICloudEventFunction<MessagePublishedData>
+    {
+        private readonly ILogger _logger;
+
+        public Function(ILogger<Function> logger) =>
+            _logger = logger;
+
+        public Task HandleAsync(CloudEvent cloudEvent, MessagePublishedData data, CancellationToken cancellationToken)
+        {
+            bool retry = false;
+            string text = data.Message?.TextData;
+
+            // Get the value of the "retry" JSON parameter, if one exists.
+            if (!string.IsNullOrEmpty(text))
+            {
+                JsonElement element = JsonSerializer.Deserialize<JsonElement>(data.Message.TextData);
+
+                retry = element.TryGetProperty("retry", out var property) &&
+                    property.ValueKind == JsonValueKind.True;
+            }
+
+            // Throwing an exception causes the execution to be retried.
```

Comment:
I don't fully understand this sample but it might just be me not having all the context. Questions:

Is this a purely Pub/Sub kind of sample?
If it is, is throwing an exception equivalent to nack?
If it is not, is throwing an exception in the context of cloud events means that the event will be reattempted?

Why the retry property? Is this something that PubSub sends to indicate what?

Reply:
I *believe* any CloudEvent will be retried. But yes, throwing an exception is equivalent to "nack" to the event pipeline (rather than straight to PubSub)

The purpose of checking retry in the JSON is just to give an idea of a function which may or may not fail.
This is ported from here: https://github.com/GoogleCloudPlatform/java-docs-samples/blob/master/functions/concepts/retry-pubsub/src/main/java/functions/RetryPubSub.java